src/datasets/arrow_dataset.py

```diff
@@ -4388,14 +4388,15 @@ def add_column(self, name: str, column: Union[list, np.array], new_fingerprint:
         })
         ```
         """
-        column_table = InMemoryTable.from_pydict({name: column})
+        ts = OptimizedTypedSequence(column, type=None, col=name)
+        column_table = InMemoryTable.from_pydict({name: ts})
         _check_column_names(self._data.column_names + column_table.column_names)
         dataset = self.flatten_indices() if self._indices is not None else self
         # Concatenate tables horizontally
         table = concat_tables([dataset._data, column_table], axis=1)
         # Update features
         info = dataset.info.copy()
-        info.features.update(Features.from_arrow_schema(column_table.schema))
+        info.features.update(Features({name: ts.get_inferred_type()}))
```

Comment:
Why not keep `Features.from_arrow_schema` here ? `get_inferred_type` instantiates an extra array to infer the types.

Reply:
`get_inferred_type` is needed for the type inference (e.g. for a sequence of PIL Images `Features.from_arrow_schema` would return `{'bytes': Value(dtype='binary', id=None), 'path': Value(dtype='string', id=None)}` instead of `Image(decode=True, id=None)`).

Reply:
Oh ok I see ! Anyway later we can cache the inferred type when `__arrow_array__` is called for the first time, this way `get_inferred_type` doesn't recreate an arrow array